pkg/antctl/raw/upgrade/apistorage/command.go

```diff
@@ -0,0 +1,260 @@
+	} else {
+		// If the user-provided name list of CRDs to upgrade is not empty, and it contains CRDs without suffix
+		// "crd.antrea.io", skip these CRDs.
+		for name := range crdNamesToUpgrade.Difference(antreaCRDNames) {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip CRD %q which is not created by Antrea.\n", name)
+		}
+		// Only upgrade the CRDs with suffix "crd.antrea.io".
+		crdNamesToUpgrade = crdNamesToUpgrade.Intersection(antreaCRDNames)
+	}
+	return crdNamesToUpgrade, nil
+}
+
+// getCRDsToUpgrade gets a list of Antrea CRDs to upgrade.
+func getCRDsToUpgrade(cmd *cobra.Command, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) []*apiextv1.CustomResourceDefinition {
+	var crdsToUpgrade []*apiextv1.CustomResourceDefinition
+	for crdName := range crdNameListToUpgrade {
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			fmt.Fprintf(cmd.ErrOrStderr(), "Failed to get CRD %q: %v.\n", crdName, err)
+			continue
+		}
+		// Skip the CRD that has only one version.
+		if len(crd.Spec.Versions) < 2 {
+			fmt.Fprintf(cmd.OutOrStdout(), "Skip upgrading CRD %q since it only has one version.\n", crdName)
+			continue
+		}
+		crdsToUpgrade = append(crdsToUpgrade, crd)
+	}
+	return crdsToUpgrade
+}
+
+// upgradeCRDObjects upgrades the existing objects of a CRD.
+func upgradeCRDObjects(cmd *cobra.Command, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: getCRDStorageVersion(crd),
+		Kind:    crd.Spec.Names.ListKind,
+	})
+	if err := k8sClient.List(context.TODO(), list); err != nil {
+		return err
+	}
+	if len(list.Items) == 0 {
+		return nil
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Upgrading %d objects of CRD %q.\n", len(list.Items), crd.Name)
+	for _, obj := range list.Items {
+		copyObj := obj
+		err := retry.OnError(wait.Backoff{
+			Duration: time.Second,
+			Steps:    5,
+		}, func(err error) bool {
+			return validateErr(err) != nil
+		}, func() error {
+			return k8sClient.Update(context.TODO(), &copyObj)
+		})
+		if validateErr(err) != nil {
+			return err
+		}
+	}
+	fmt.Fprintf(cmd.OutOrStdout(), "Successfully upgraded %d objects of CRD %q.\n", len(list.Items), crd.Name)
+	return nil
+}
+
+// patchCRDStoredVersions patches status.storedVersion of a CRD.
+func patchCRDStoredVersions(k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
+	freshCRD := &apiextv1.CustomResourceDefinition{}
+	if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crd.Name}, freshCRD); err != nil {
+		return err
+	}
+
+	expectedStorageVersion := getCRDStorageVersion(crd)
+	// This ensures that the storage version is not changed during the upgrade.
+	if getCRDStorageVersion(freshCRD) != expectedStorageVersion {
+		return newUnexpectedChangeError(crd)
+	}
```

Comment:
This is perhaps unnecessary because:
1. Even if you check it, it's still possible that the CRD is updated after you get the "fresh" version.
2. We should leverage the resourceVersion to let apiserver fail the update request when the CRD is updated after we get it.

Reply:
> 1. Even if you check it, it's still possible that the CRD is updated after you get the "fresh" version.

I got this and removed related code.

> 2. We should leverage the resourceVersion to let apiserver fail the update request when the CRD is updated after we get it.

Could you give more details about this?

Reply:
The code seems not removed. What I mean is, we could assume the CRD is not updated during the upgrade. It could just update the `StoredVersions` to the storage version in the CRD's spec. If the CRD is updated during the upgrade, the update will fail because the resourceVersion doesn't match. ResourceVersion is used for this situation, preventing one client overriding another's write accidently.

If the update request fails with Conflict error, it means the CRD might have been updated in some way by users. We could just return error and abort the process.